sklearn/utils/testing.py

```diff
@@ -393,6 +403,177 @@ def assert_raise_message(exception, message, function, *args, **kwargs):
         exception, re.escape(message), function, *args, **kwargs)
 
 
+def assert_array_not_equal(x, y, err_msg="The arrays are equal", verbose=True):
+    """Raises an AssertionError if two array-like objects are equal."""
+    try:
+        assert_array_equal(x, y, verbose=verbose)
+    except AssertionError:
+        return
+    raise AssertionError(err_msg)
+
+
+def _assert_same_model_method(method, X, estimator1, estimator2, msg=None):
+    if hasattr(estimator1, method):
+        m = '%r\n\nhas %s, but\n\n%r\n\ndoes not' % (estimator1,
+                                                     method,
+                                                     estimator2)
+
+        if not hasattr(estimator2, method):
+            raise AttributeError(m)
+
+        # Check if the method(X) returns the same for both models.
+        res1 = getattr(estimator1, method)(X)
+        res2 = getattr(estimator2, method)(X)
+        same_model = (res1.shape == res2.shape) and np.allclose(res1, res2)
+
+        if msg is None:
+            msg = ("Models are not equal. \n\n%s method returned"
+                   " different results:\n\n%r\n\n and\n\n%r\n\n for :\n\n%r"
+                   "\n\nand :\n\n%r" % (method, res1, res2,
+                                        estimator1, estimator2))
+
+        assert same_model, msg
+
+
+def _not_same_model_method(method, X, estimator1, estimator2):
+    """Return True if both estimators return same result for given method"""
+    if hasattr(estimator1, method):
+        m = '%r\n\nhas %s, but\n\n%r\n\ndoes not' % (estimator1,
+                                                     method,
+                                                     estimator2)
+
+        if not hasattr(estimator2, method):
+            raise AttributeError(m)
+
+        # Check if the method(X) returns the same result for both models.
+        res1 = getattr(estimator1, method)(X)
+        res2 = getattr(estimator2, method)(X)
+        not_same_model = (res1.shape != res2.shape) or not np.allclose(res1,
+                                                                       res2)
+
+        # NOTE Do not assert here. This is to facilitate checking using any()
+        # of the results of multiple checks.
+        return not_same_model
+    return False  # if the method does not exist.
+
+
+def assert_same_model(X, estimator1, estimator2, msg=None):
+    """Helper function to check if models are same"""
+    _assert_same_model_method('predict', X, estimator1, estimator2, msg)
+    _assert_same_model_method('transform', X, estimator1, estimator2, msg)
+    _assert_same_model_method('decision_function',
+                              X, estimator1, estimator2, msg)
+    _assert_same_model_method('predict_proba',
+                              X, estimator1, estimator2, msg)
+
+    assert_fitted_attributes_equal(estimator1, estimator2)
+
+
+def assert_not_same_model(X, estimator1, estimator2, msg=None):
```

Comment:
+1

This would simplify things a lot. We could drop `assert_fitted_attributes_not_equal` entirely as it is a YAGNI and collapse the body of `_attributes_equal` directly into `assert_fitted_attributes_equal` without having to deal with the negative case at all. That would make it possible to use `assert_array_almost_equal` directly to check values of array attributes.